drivers/soundwire/master.c

```diff
@@ -78,90 +67,82 @@ struct sdw_master_device
 		goto device_register_err;
 	}
 
-	if (link_ops && link_ops->add) {
-		ret = link_ops->add(md, pdata);
+	/* add shortcuts to improve code readability/compactness */
+	md->bus = bus;
+	bus->dev = &md->dev;
+
+	if (bus->link_ops && bus->link_ops->add) {
+		ret = bus->link_ops->add(bus, bus->pdata);
 		if (ret < 0) {
-			dev_err(&md->dev, "link_ops add callback failed: %d\n", ret);
+			dev_err(&md->dev,
+				"link_ops add callback failed: %d\n", ret);
 			goto link_add_err;
 		}
 	}
 
-	return md;
+	return ret;
 
 link_add_err:
 	device_unregister(&md->dev);
 device_register_err:
-	return ERR_PTR(ret);
+	return ret;
 }
-EXPORT_SYMBOL_GPL(sdw_master_device_add);
 
 /**
  * sdw_master_device_del() - delete a Linux Master Device representation.
- * @md: the master device
+ * @bus: bus handle
  *
- * This function is the dual of sdw_master_device_add(), itreleases
- * all link-specific resources and unregisters the device.
+ * This function is the dual of sdw_master_device_add()
  */
-int sdw_master_device_del(struct sdw_master_device *md)
+int sdw_master_device_del(struct sdw_bus *bus)
 {
-	int ret = 0;
+	int ret;
 
-	if (md && md->link_ops && md->link_ops->del) {
-		ret = md->link_ops->del(md);
-		if (ret < 0) {
-			dev_err(&md->dev, "link_ops del callback failed: %d\n",
+	if (!bus)
+		return -EINVAL;
+
+	if (bus->link_ops && bus->link_ops->del) {
+		ret = bus->link_ops->del(bus);
+		if (ret < 0)
+			dev_err(bus->dev,
+				"link_ops del callback failed: %d\n",
 				ret);
-			return ret;
```

Comment:
So, should we  return ret here?

Reply:
yes we should - though it doesn't really matter because sdw_bus_master_delete() returns void.

Reply:
actually no, sorry. If the link_ops fails, we still want unregister the device.